### PR TITLE
refactor: make `GameState` opaque

### PR DESCRIPTION
**Pull request by JonathanStarup**
Fixes part of #17.

I don't know if this makes sense or not. I'm not used to thinking like this in flix. opaque forces you to separate functions into internal vs external. I suppose Engine is internal to GameState right?

### CODE REVIEW

**Comment by JonathanStarup** [comment]
According to the Official Data Representation Matrix™
|                         | ADT | Records | ADT with record | opaque ADT | opaque ADT with record |
| :---------------------- | :-: | :-----: | :-------------: | :---: | :---: |
| X(Y) means External(Internal) | | | | | |
| Pattern matching          | ✔ | ❌ | ❌ |  ❌(✔) | ❌(❌) |
| Direct data access         | get/set(2*field defs)| ✔ | unwrap(1 def) | ❌(get/set) | ❌(unwrap) |
| Stable under extension    | ❌ | ❌ | ❌ | ✔(❌) | ✔(❌) |
| Works with typeclasses  | ✔ | ❌ | ✔ | ✔ | ✔ |

We don't care about pattern matching on the game state (too complex) so opaque ADT with record is the best choice (we don't have to write getters/setters

**Comment by mlutze** [comment]
My concern with making it a record is that we lose a little bit of compile-time protections. For example, say we add another map `PlayerId -> X` to our game state. We have the invariant that the keys of each `PlayerId -> X` map are the same. The `removePlayer` function needs to remove the playerId from this map as well, but with a record it is easy to forget about this case.

That said, I think that actually points to a flaw in our model. If we have just one map internally `PlayerId -> PlayerInfo` then we get that invariant for free.

tl;dr: Seems reasonable, but we should change the game state structure.

**Comment by mlutze** [comment]
I'm not sure that `Engine` is internal to `GameState` though. How hard is it to separate them and use functions to modify `GameState`?

**Comment by JonathanStarup** [comment]
> I'm not sure that `Engine` is internal to `GameState` though. How hard is it to separate them and use functions to modify `GameState`?

Isn't it inherent to gamestate that it can run? and also, if we restructure gamestate, shouldn't engine also be restructured so that its efficient as possible?

**Comment by JonathanStarup** [comment]
Another annoying thing: I want to use namespaces to have short names. I don't want to have mkGameState but instead GameState.init . Its very annoying without namespace aliasing though because you can only write `Flixball/Core/GameState.init`

**Comment by mlutze** [comment]
> > I'm not sure that `Engine` is internal to `GameState` though. How hard is it to separate them and use functions to modify `GameState`?
> 
> Isn't it inherent to gamestate that it can run? and also, if we restructure gamestate, shouldn't engine also be restructured so that its efficient as possible?

I'm thinking of `GameState` as just a data holder, whereas other things use it and manipulate it. Maybe I have too much of an OOP mindset?



> Another annoying thing: I want to use namespaces to have short names. I don't want to have mkGameState but instead GameState.init . Its very annoying without namespace aliasing though because you can only write `Flixball/Core/GameState.init`

Yes, I would like the same thing. We really need this feature.

**Comment by JonathanStarup** [comment]
> I'm thinking of GameState as just a data holder, whereas other things use it and manipulate it.

@mlutze in that mindset, it doesn't really make sense to make it opaque. This also wouldn't hurt much since its not "exposed" to the ais. Should it just be non-opaque? I can make some getters to make field changes more unnoticable. Maybe we could even just have GameState be an alias of a record type?

**Comment by mlutze** [comment]
I think the record idea could work. That way we can change the game state's fields without needing to modify everything that uses it.

**Comment by mlutze** [comment]
Then maybe `PlayerInfo` is an alias to a record type as well, so we can add fields.

**Comment by JonathanStarup** [comment]
> I think the record idea could work. That way we can change the game state's fields without needing to modify everything that uses it.

This is true except when building a new value of the type, but then again, most things would probably use a new() method and otherwise just update fields

Another downside is that you cannot have redundancy in the record because fields are just public. So if you ever want redundancy for performance reasons you have to be very careful

**Comment by mlutze** [comment]
> > I think the record idea could work. That way we can change the game state's fields without needing to modify everything that uses it.
> 
> This is true except when building a new value of the type, but then again, most things would probably use a new() method and otherwise just update fields
> 
> Another downside is that you cannot have redundancy in the record because fields are just public. So if you ever want redundancy for performance reasons you have to be very careful.
> 
> The more I think about this the more I begin to look back to good old java programming. private fields public getters (opaque enum in this context). Maybe there is no reason to invent something new, it's just frustrating that these fun tools (records, adt's) seem to have issues with extensibility.
> 
> It seems to me that you should
> 
>     * use ADT when its data is essential to its identity (no future extension) and it does not need validity checking (redundancy, integrity)
> 
>     * use Records if you're working with a "data blob" that might gain more information over time (only extension) and it does not need validity checking (redundancy, integrity)
> 
>     * Otherwise use an opaque enum
> 
> 
> So in conclusion, I guess I'm agreeing with your original opaque idea while still having engine be external. This just needs a handful nice operations on gamestate

And we already have internal redundancy between the player's position in the map and in the board! (I think)

**Comment by mlutze** [comment]
I can't do a thumbs up on mobile but ( ^ ^)-b

**Comment by JonathanStarup** [comment]
Continues in #96
